File: 49_Group Anagrams/Solution.py
from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs):
        # Create a dictionary to store groups of anagrams
        anagram_map = defaultdict(list)
        
        # Iterate through each string in the input list
        for s in strs:
            # Sort the string and use it as a key
            sorted_str = ''.join(sorted(s))
            # Add the original string to the corresponding list
            anagram_map[sorted_str].append(s)
        
        # Return the grouped anagrams as a list of lists
        return list(anagram_map.values())

# Example usage
if __name__ == "__main__":
    solution = Solution()
    
    # Example 1
    strs1 = ["eat", "tea", "tan", "ate", "nat", "bat"]
    print(solution.groupAnagrams(strs1))
    # Output: [['eat', 'tea', 'ate'], ['tan', 'nat'], ['bat']]
    
    # Example 2
    strs2 = [""]
    print(solution.groupAnagrams(strs2))
    # Output: [['']]
    
    # Example 3
    strs3 = ["a"]
    print(solution.groupAnagrams(strs3))
    # Output: [['a']]

File: 49_Group Anagrams/Solution.java
import java.util.*;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        // Create a map to store groups of anagrams
        Map<String, List<String>> map = new HashMap<>();

        // Iterate through each string in the input array
        for (String s : strs) {
            // Convert the string to a character array and sort it
            char[] charArray = s.toCharArray();
            Arrays.sort(charArray);
            String sorted = new String(charArray);

            // If the sorted string is not in the map, add it with a new list
            if (!map.containsKey(sorted)) {
                map.put(sorted, new ArrayList<>());
            }
            // Add the original string to the corresponding list
            map.get(sorted).add(s);
        }

        // Return the grouped anagrams as a list of lists
        return new ArrayList<>(map.values());
    }

    // Main method for testing
    public static void main(String[] args) {
        Solution solution = new Solution();
        
        // Example 1
        String[] strs1 = {"eat", "tea", "tan", "ate", "nat", "bat"};
        System.out.println(solution.groupAnagrams(strs1)); 
        // Output: [["bat"],["nat","tan"],["ate","eat","tea"]]

        // Example 2
        String[] strs2 = {""};
        System.out.println(solution.groupAnagrams(strs2)); 
        // Output: [[""]]

        // Example 3
        String[] strs3 = {"a"};
        System.out.println(solution.groupAnagrams(strs3)); 
        // Output: [["a"]]
    }
}